### PR TITLE
Best score callback

### LINKED ISSUE #2849: Improve Documentation for ValidationCallback to Assist with Hyperparameter Tuning
## 🚀 Feature

I took a closer look at ValidationCallback in deepchem/models/callbacks.py and I don't think any more code needs to be written. There is already an attribute named self._best_score and an argument named save_on_minimum which allows choosing minimize or maximize. 

I have a few ideas: 1) Add a function to ValidationCallback called GetBestScore() or something. It may honestly not be necessary as like I said, self._best_score is already there. 2) Edit the dc.hyper.GridHyperparamOpt tuner to use the best validation score. I think it currently evaluates the FINAL model on the validation set instead of the BEST model but please correct me if I'm wrong. Then document this change clearly so others who wish to create a new hyperparameter optimizer will have an example to follow.

I think as a follow up, actual early stopping where the model saves the best model and stops if it doesn't improve after n steps would be a really nice feature to have, although I'm sure it'd be very complicated to implement across so many models. 

## Motivation

Allowing users to access the best validation score for the purpose of early stopping is very important for hyperparameter optimization. This feature is already there but the question is how do we make it easier to use/develop with. 

## Additional context

### PR DESCRIPTION
**Pull request by TheRealSalmon**
# Pull Request Template

## Description

Fix #2849

Rearranged some code to make it so ValidationCallback always updates self._best_score. Previously, it was only doing so if self.save_dir != None. I also added a getter called get_best_score(). Finally I updated/added some docstrings and added a unit test.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by peastman** [comment]
In case you didn't already know, you don't need to close a PR to make changes to it.  Just push changes to the branch and they'll get included automatically.

**Comment by TheRealSalmon** [comment]
I removed the unused tensorflow import. Is there anything else I need to do to finish this PR?

**Comment by rbharath** [comment]
@TheRealSalmon Can you make it to one of the developer calls this week? I've authorized the CI run on the PR and it might be easiest to finish the a review on the call

**Comment by TheRealSalmon** [comment]
@rbharath yes see you Thursday evening at office hours and dev call.

**Comment by rbharath** [comment]
It looks like the test is failing: https://github.com/deepchem/deepchem/runs/5328745202?check_suite_focus=true

Can you run tests locally? That might help as you're debugging

**Comment by TheRealSalmon** [comment]
@rbharath I fixed the test to make more sense and to pass. Could you start the test before our meeting Thursday?

**Comment by TheRealSalmon** [comment]
@rbharath I the CI testing has passed the ~30 threshold. Is it time to get it merged in?

**Comment by rbharath** [comment]
Yes this looks good now! Congrats on your second merged PR :)
